### PR TITLE
[mlir][emitc] Add 'emitc.switch' op to the dialect

### PR DESCRIPTION
**Pull request by EtoAndruwa**
This PR is continuation of the [previous one](https://github.com/llvm/llvm-project/pull/101478).  As a result, the `emitc::SwitchOp` op was developed inspired by `scf::IndexSwitchOp`.

Main points of PR:

- Added the `emitc::SwitchOp` op  to the EmitC dialect + CppEmitter
- Corresponding tests were added
- Conversion from the SCF dialect to the EmitC dialect for the op

### CODE REVIEW

File: mlir/test/Target/Cpp/switch.mlir
Comment (mgehre-amd) [inline mlir/test/Target/Cpp/switch.mlir:87]: Could you please add a test case that shows a switch without a default label?

File: mlir/test/Target/Cpp/invalid.mlir
Comment (mgehre-amd) [inline mlir/test/Target/Cpp/invalid.mlir:216]: Could you please add a test case with duplicated case values?

File: mlir/test/Target/Cpp/switch.mlir
Comment (mgehre-amd) [inline mlir/test/Target/Cpp/switch.mlir:129]: The documentation for emitc.yield says `If the parent operation defines no values, then the `emitc.yield` may be left out in the custom syntax and the builders will insert one implicitly.`
Does that work here? If that would work, I think it would be easy to confuse with fall-through, which this PR doesn't support.

File: mlir/lib/Target/Cpp/TranslateToCpp.cpp
Comment (mgehre-amd) [inline mlir/lib/Target/Cpp/TranslateToCpp.cpp:455]: Would `drop_end` https://github.com/Xilinx/llvm-project/blob/729aef473d0efc175db1fa56e2215c634c991bd6/llvm/include/llvm/ADT/STLExtras.h#L336 work here?

File: mlir/lib/Target/Cpp/TranslateToCpp.cpp
Comment (mgehre-amd) [inline mlir/lib/Target/Cpp/TranslateToCpp.cpp:472]: Can we do a range-based for loop over `llvm::zip(getCases(), getCaseBlocks())` instead of manual iterator handling?

File: mlir/lib/Target/Cpp/TranslateToCpp.cpp
Comment (mgehre-amd) [inline mlir/lib/Target/Cpp/TranslateToCpp.cpp:452]:
```suggestion
static LogicalResult emitSwitchCase(CppEmitter &emitter,
```
due to writing the `break;` at the end of the function, the name can be more specific.

Comment (EtoAndruwa) [reply mlir/lib/Target/Cpp/TranslateToCpp.cpp:452]: Agree, it looks much better.

File: mlir/test/Target/Cpp/invalid.mlir
Comment (EtoAndruwa) [reply mlir/test/Target/Cpp/invalid.mlir:216]: Added.

File: mlir/test/Target/Cpp/switch.mlir
Comment (EtoAndruwa) [reply mlir/test/Target/Cpp/switch.mlir:87]: Added.

File: mlir/lib/Target/Cpp/TranslateToCpp.cpp
Comment (EtoAndruwa) [reply mlir/lib/Target/Cpp/TranslateToCpp.cpp:472]: I will refactor this using `llvm::zip(getCases(), getCaseBlocks())`.

File: mlir/test/Target/Cpp/switch.mlir
Comment (EtoAndruwa) [reply mlir/test/Target/Cpp/switch.mlir:129]: Right now, you must explicitly yield each case region and the default one. We can support the implicit addition of `emitc::YieldOp` to the end of each region in the parser, as it is done in `emitc::IfOp`'s parser with `IfOp::ensureTerminator(...)`.

Comment (mgehre-amd) [reply mlir/test/Target/Cpp/switch.mlir:129]: Ok, then I think it's better to continue requiring explicit `yields` for now so users don't get confused about a missing yield and think that this is doing fall-through.

File: mlir/lib/Dialect/EmitC/IR/EmitC.cpp
Comment (mgehre-amd) [reply mlir/lib/Dialect/EmitC/IR/EmitC.cpp:135]: It's fine for users to use the IfOp for boolean conditions, but that doesn't mean we should disallow using the SwitchOp. Those restrictions can make writing generic lowerings more complicated because they would need to special case for `i1`.

The `emitc` IR is mostly a syntactic representation of the C++ code. So I'd think that emitc.switch should allow all integral types as condition. C++ doesn't require to explicitly cast to i32 (because it does it implicitly), so we shouldn't require the cast in the emitc IR either.
(It's also not so clear why e.g. i16 and i64 would be different than i8 with respect to resolveOperand).

**Comment by EtoAndruwa** [comment]
From now on, I will commit all changes with separate commits and then rebase after approval.

**Comment by mgehre-amd** [comment]
> From now on, I will commit all changes with separate commits and then rebase after approval. It will be more convenient.

Sounds good. You don't even need to rebase manually; the `Squash and merge` button on the PR will automatically combine all commits into a single one.

File: mlir/lib/Dialect/EmitC/IR/EmitC.cpp
Comment (EtoAndruwa) [reply mlir/lib/Dialect/EmitC/IR/EmitC.cpp:135]: > The emitc IR is mostly a syntactic representation of the C++ code. So I'd think that emitc.switch should allow all integral types as condition.

I agree with you. We will support all [types](https://en.cppreference.com/w/cpp/language/switch) for the condition. I will upload the respective patch.

File: mlir/test/Target/Cpp/switch.mlir
Comment (EtoAndruwa) [reply mlir/test/Target/Cpp/switch.mlir:129]: > Ok, then I think it's better to continue requiring explicit yields for now so users don't get confused about a missing yield and think that this is doing fall-through.

I will add `requiring explicit yields` to the description of the op in the `.td` file. Can I resolve conversation?

Comment (mgehre-amd) [reply mlir/test/Target/Cpp/switch.mlir:129]: yes, that sounds good

File: mlir/lib/Dialect/EmitC/IR/EmitC.cpp
Comment (EtoAndruwa) [reply mlir/lib/Dialect/EmitC/IR/EmitC.cpp:135]: Just found that `i1` `true` converts to `-1` in `printOperation(CppEmitter &emitter, emitc::SwitchOp switchOp)`. I will fix it with a separate patch.

**Comment by aniragil** [comment]
>     * Conversion from the SCF dialect to the EmitC dialect will be supported after the approvement this/adjusted version of the `emitc::SwitchOp` op

Any reason why lowering from `scf.index_switch` isn't part of this PR (as done for `emitc.if` and `emitc.for`)?
Without it, the op is dead upstream (i.e. not reachable from any higher dialect). In addition, we can't validate that its functionality & implementation fit its upstream-designated purpose, which is to be the lowering of `scf.index_switch`.

**Review by aniragil — CHANGES_REQUESTED** [review]

**Comment by EtoAndruwa** [comment]
> >     * Conversion from the SCF dialect to the EmitC dialect will be supported after the approvement this/adjusted version of the `emitc::SwitchOp` op
> 
> Any reason why lowering from `scf.index_switch` isn't part of this PR (as done for `emitc.if` and `emitc.for`)?
> Without it, the op is dead upstream (i.e. not reachable from any higher dialect). In addition, we can't validate that its functionality & implementation fit its upstream-designated purpose, which is to be the lowering of `scf.index_switch`.

Hello! The reason why the conversion is not a part of this PR is that we wanted to get feedback on the implementation of the op as quickly as possible and make adjustments. Conversion **will be added** after the op's implementation is approved. It will be included as part of this PR.

File: mlir/include/mlir/Dialect/EmitC/IR/EmitC.td
Comment (marbre) [inline mlir/include/mlir/Dialect/EmitC/IR/EmitC.td:1327]: What could help here would be a short example that shows the emitted code for one of the MLIR examples.

File: mlir/lib/Dialect/EmitC/IR/EmitC.cpp
Comment (marbre) [inline mlir/lib/Dialect/EmitC/IR/EmitC.cpp:135]:
```suggestion
```
Unused variable, which the builtbots are complaining about.

File: mlir/lib/Dialect/EmitC/IR/EmitC.cpp
Comment (marbre) [reply mlir/lib/Dialect/EmitC/IR/EmitC.cpp:135]: Just to clarify, please add to this patch (no need to rebase, separate commit is fine) but don't address in a follow up PR. I would refer this PR as one single patch, as it needs to gets squashed at the end.

File: mlir/test/Target/Cpp/switch.mlir
Comment (marbre) [inline mlir/test/Target/Cpp/switch.mlir:1]: Can please take a look if anything breaks when passing `-declare-variables-at-top` and eventually add a test, see e.g.:

https://github.com/llvm/llvm-project/blob/7e175b307e12a616b287f58af0fbfd8d9164e364/mlir/test/Target/Cpp/for.mlir#L1-L2

**Review by marbre — CHANGES_REQUESTED** [review]
As @aniragil mentioned it would be great to see the conversion as part of this PR as well to see if the op provides what we need or if something within the op is broken :)

**Comment by EtoAndruwa** [comment]
> As @aniragil mentioned it would be great to see the conversion as part of this PR as well to see if the op provides what we need or if something within the op is broken :)

I will try to add the conversion today.

File: mlir/lib/Dialect/EmitC/IR/EmitC.cpp
Comment (EtoAndruwa) [reply mlir/lib/Dialect/EmitC/IR/EmitC.cpp:135]: > The emitc IR is mostly a syntactic representation of the C++ code. So I'd think that emitc.switch should allow all integral types as condition.
 
To be precise, we want to allow all integral types, opaque and index? Cases for the switch must always be integers (arg will be implicitly casted to i64). And there is no need to add the logic for parsing something like this: `case true:`, `case ENUM::...:`. Am I right?

Comment (simon-camp) [reply mlir/lib/Dialect/EmitC/IR/EmitC.cpp:135]: Would it be hard to switch to a plain ArrayAttr and parse or verify Integer, Index or Opaque attributes for the cases?

Comment (EtoAndruwa) [reply mlir/lib/Dialect/EmitC/IR/EmitC.cpp]: Right now we are supporting index, opaque and integral types as arguments for the switch operator. Tests for all types we added

Comment (EtoAndruwa) [reply mlir/lib/Dialect/EmitC/IR/EmitC.cpp]: Can I resolve conversation?

File: mlir/include/mlir/Dialect/EmitC/IR/EmitC.td
Comment (marbre) [inline mlir/include/mlir/Dialect/EmitC/IR/EmitC.td:1193]: With this change you need to update https://github.com/llvm/llvm-project/blob/b9793a6f04b3ab1ff9ed86a675c93e957ec52f51/mlir/test/Dialect/EmitC/invalid_ops.mlir#L230 accordingly. The builtbot fails, please see the logs!

**Review by marbre — CHANGES_REQUESTED** [review]

File: mlir/include/mlir/Dialect/EmitC/IR/EmitC.td
Comment (aniragil) [inline mlir/include/mlir/Dialect/EmitC/IR/EmitC.td:1330]: Examples section seems overly detailed (cf. to examples in other EmitC ops):

- Selector's type doesn't imply significantly different semantics/output, so demonstrating a single type seems enough
- The `--declare-variables-at-top` flag affects other ops in a similar manner (and isn't demonstrated for them)
- Showing the op itself (without the function context) seems sufficient
- A single variable + call in the default section seems enough

Interested users can find more detailed examples in the dialect's tests.

**Comment by EtoAndruwa** [comment]
Just added conversion from `scf::index_switch` to `emitc::switch`. Will wait for your review @aniragil, @marbre, @mgehre-amd.

**Review by aniragil — CHANGES_REQUESTED** [review]

File: mlir/lib/Conversion/SCFToEmitC/SCFToEmitC.cpp
Comment (aniragil) [inline mlir/lib/Conversion/SCFToEmitC/SCFToEmitC.cpp:209]: Can we refactor this function and its equivalent from `scf.if` lowering into a static function (similar to `lowerYield()`)?

**Review by aniragil — CHANGES_REQUESTED** [review]

File: mlir/include/mlir/Dialect/EmitC/IR/EmitC.td
Comment (marbre) [inline mlir/include/mlir/Dialect/EmitC/IR/EmitC.td:58]: If there is no good reason why pointer wide types should not be used, this should be dropped in favor of using the existing `IntegerIndexOrOpaqueType` which is referring to `isIntegerIndexOrOpaqueType`.

Comment (EtoAndruwa) [reply mlir/include/mlir/Dialect/EmitC/IR/EmitC.td:58]: We will support them too. Must I add respective test cases?

File: mlir/lib/Dialect/EmitC/IR/EmitC.cpp
Comment (aniragil) [inline mlir/lib/Dialect/EmitC/IR/EmitC.cpp:1183]: nit: please combine these to a single statement (reduce line noise)

File: mlir/lib/Dialect/EmitC/IR/EmitC.cpp
Comment (aniragil) [inline mlir/lib/Dialect/EmitC/IR/EmitC.cpp:1191]: nit: please combine these to a single statement (reduce line noise)

File: mlir/lib/Dialect/EmitC/IR/EmitC.cpp
Comment (aniragil) [inline mlir/lib/Dialect/EmitC/IR/EmitC.cpp:1210]: nit: please add an empty line above.

**Review by aniragil — CHANGES_REQUESTED** [review]

File: mlir/lib/Target/Cpp/TranslateToCpp.cpp
Comment (simon-camp) [inline mlir/lib/Target/Cpp/TranslateToCpp.cpp:470]: Do we want to have parentheses around the cases?

Comment (EtoAndruwa) [reply mlir/lib/Target/Cpp/TranslateToCpp.cpp:470]: Cases are `int64_t`, so I will delete them.

**Comment by EtoAndruwa** [comment]
I will upload new testes and upload new one for updates.

**Review by simon-camp — CHANGES_REQUESTED** [review]
Overall this looks good, just two things that should be fixed before merging.

File: mlir/lib/Dialect/EmitC/IR/EmitC.cpp
Comment (EtoAndruwa) [reply mlir/lib/Dialect/EmitC/IR/EmitC.cpp:1184]: What is a reason for printing `:` after the `default` keyword? This is refactored version of `IndexSwitchOp::print` function, where is no colon after the `default`.

File: mlir/include/mlir/Dialect/EmitC/IR/EmitC.td
Comment (EtoAndruwa) [reply mlir/include/mlir/Dialect/EmitC/IR/EmitC.td]: Added tests.

File: mlir/lib/Dialect/EmitC/IR/EmitC.cpp
Comment (simon-camp) [reply mlir/lib/Dialect/EmitC/IR/EmitC.cpp:1184]: I've checked out the code and ran a few tests.

You are currently expecting a colon after the default keyword in the parser. Can you then please remove  the colons from `parseSwitchCases` and `printSwitchCases`, so these functions are a 1:1 copy of the scf index switch version.

Additionally you are parsing `: type` after the argument, but you are not printing that, so the custom syntax does not roundtrip.

Therefore I suggest switching to custom assembly in tablegen similar to the scf dialect:
```
let assemblyFormat = [{
    $arg `:` type($arg) attr-dict custom<SwitchCases>($cases, $caseRegions) `\n`
    `` `default` $defaultRegion
  }];
```

You can then delete `SwitchOp::parse` and `SwitchOp::print` as these will be auto generated.

Additionally you need to add a valid switch op in `test/Dialect/EmitC/ops.mlir`; this file tests roundtripping of the custom assembly format.

File: mlir/include/mlir/Dialect/EmitC/IR/EmitC.td
Comment (simon-camp) [inline mlir/include/mlir/Dialect/EmitC/IR/EmitC.td:1340]: nit: Remove the colons here.

File: mlir/include/mlir/Dialect/EmitC/IR/EmitC.td
Comment (simon-camp) [inline mlir/include/mlir/Dialect/EmitC/IR/EmitC.td:1351]: nit: Remove the parentheses around the case values.

**Review by simon-camp — APPROVED** [review]
LGTM, just update the op description to the new syntax please.

**Review by simon-camp — APPROVED** [review]

**Comment by EtoAndruwa** [comment]
Hello, @marbre and @aniragil! Can you approve changes?

**Review by aniragil — APPROVED** [review]
LGTM!, thanks @EtoAndruwa!

**Comment by EtoAndruwa** [comment]
Hello, @mgehre-amd! Can you please approve changes?

**Review by mgehre-amd — APPROVED** [review]

File: mlir/lib/Conversion/SCFToEmitC/SCFToEmitC.cpp
Comment (marbre) [inline mlir/lib/Conversion/SCFToEmitC/SCFToEmitC.cpp:98]: NIT:
```suggestion
// emitc::if/emitc::switch region. The contents of the lowering region is
```

File: mlir/include/mlir/Dialect/EmitC/IR/EmitC.td
Comment (marbre) [inline mlir/include/mlir/Dialect/EmitC/IR/EmitC.td:1345]:
```suggestion
    ```
    ```c++
    // Code emitted for the operations above.
```
With this you get the correct syntax highlighting for the generated C++ code.

**Review by marbre — APPROVED** [review]
I unfortunately had only time for a quick scan of the current revision. Just a few small nits. Should be fine if it gets merged afterwards.

**Review by marbre — APPROVED** [review]
LGTM! Thanks for working on this and taking care of all the review comments.

**Comment by EtoAndruwa** [comment]
Hello, @marbre! Can you please merge this after more detailed check?

**Comment by EtoAndruwa** [comment]
> I unfortunately had only time for a quick scan of the current revision.

Hello, @marbre! Can you please merge this after more detailed check?

**Comment by marbre** [comment]
Just wanted to hit the merge button and noticed you're using a private email which the bot should have caught before, see
https://github.com/llvm/llvm-project/blob/66878ff69293c4008707261592d448d5896239e7/.github/workflows/email-check.yaml#L33-L35

**Comment by EtoAndruwa** [comment]
> Just wanted to hit the merge button and noticed you're using a private email which the bot should have caught before, see
> 
> https://github.com/llvm/llvm-project/blob/66878ff69293c4008707261592d448d5896239e7/.github/workflows/email-check.yaml#L33-L35

Fixed!

**Pull request merged**
